### PR TITLE
Implement .so check by name

### PR DESCRIPTION
**Pull request by MateuszMalisz**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

What does the PR accomplish, why was it needed?

When updating a package, the version of the provided static object files (.so files) can change. However, since no other package has been updated, they will be rebuilt locally and linked against a newer version of a library, but never published. This means that a fix might not get deployed to systems which are running packages that depend on the older version of the library.

To counter this I have implemented some targets to be run after a build process. The targets will provide list of files that prompt a version update of dependent packages, as well as a list of packages that need to be updated. See the modified documentation for more information on usage.

###### Change Log  

- Add an option to generate a list of built packages
- Add an option to check whether new packages have introduced any new .so files
- Move build summary generation to analysis.mk

###### Does this affect the toolchain?  

**NO**

###### Test Methodology

- Local run

### CODE REVIEW

File: toolkit/Makefile [hunk 1]
```diff
@@ -178,6 +178,10 @@ include $(SCRIPTS_DIR)/imggen.mk
 #   package-toolkit, clean-package-toolkit
 include $(SCRIPTS_DIR)/toolkit.mk
 
+# Analyze build process and its artifacts with:
+#
```
Comment (MateuszMalisz) [inline toolkit/Makefile:182]: add targets

**Review by christopherco — COMMENTED** [review]
In the PR description, please add some information on how to use the new analysis.mk targets (and/or add this info into the docs area).

Also there is a blank "test" file at the root folder.

File: toolkit/docs/how_it_works/6_analysis.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  - [Artifacts](#Artifacts)
+
+## Built packages list
+A `built-packages-summary` target checks the build logs and creates a file with a list of packages that have been built locally (which might be different from a list of packages present, e.g. downloaded toolchain packages). The file is used later for sodiff analysis.
+
+### Fake list
+If no packages has been built, a target `fake-built-packages-list` can be run to generate a list as if all the packages in the rpm output folder were built locally.
+
+## sodiff
+sodiff is a process which looks for new versions of .SO files and provides a list of packages that need to be rebuilt due to the change.
```
Comment (anphel31) [inline toolkit/docs/how_it_works/6_analysis.md:17]: I think this should be lowercase (".so")

File: toolkit/docs/how_it_works/6_analysis.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+To provide a list of packages that need to be rebuilt, simply run `sodiff-check` target. The target will fail if no packages has been built. In that case, one can run `fake-built-packages-list` target before making a sodiff check. This will allow to analyze all RPMs except just the locally built ones.
+
+### Implementation
+
+### Artifacts
+The artifacts are available in the SODIFF_OUTPUT_FOLDER, which is built/sodiff by default.
```
Comment (anphel31) [inline toolkit/docs/how_it_works/6_analysis.md:24]: Should this be "build/sodiff" ?

File: toolkit/docs/how_it_works/6_analysis.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+The results of sodiff check are:
+    - summary.txt file, containing a list of source rpm names (without the extension part) of packages that need to be rebuilt due to .so version change,
```
Comment (anphel31) [inline toolkit/docs/how_it_works/6_analysis.md:26]: Can you highlight the specific files (ex: `summary.txt` , `sodiff.log` )

File: toolkit/Makefile [hunk 5]
```diff
@@ -178,6 +178,11 @@ include $(SCRIPTS_DIR)/imggen.mk
 #   package-toolkit, clean-package-toolkit
 include $(SCRIPTS_DIR)/toolkit.mk
 
+# Analyze build process and its artifacts with:
+# built-packages-summary build-summary fake-built-packages-list abidiff-name-check
```
Comment (MateuszMalisz) [inline toolkit/Makefile:182]: Update target names

File: toolkit/scripts/analysis.mk [hunk 6]
```diff
@@ -0,0 +1,58 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+# Contains:
+#	- Generate list of built packages
+#	- Run check for ABI changes of built packages.
+#       - Run check for .so files version change of built packages.
```
Comment (MateuszMalisz) [inline toolkit/scripts/analysis.mk:7]: Align

**Review by christopherco — APPROVED** [review]
Thanks for making the requested documentation changes!

File: toolkit/scripts/sodiff/sodiff.repo [hunk 7]
```diff
@@ -0,0 +1,19 @@
+[mariner-official-base]
```
Comment (jslobodzian) [inline toolkit/scripts/sodiff/sodiff.repo:1]: Please use the files from the toolkit or from the mariner-repos folder rather than duplicating it.

**Review by jslobodzian — APPROVED** [review]

**Review by anphel31 — APPROVED** [review]

**Pull request merged**